provisionerd/provisionerd_test.go

```diff
@@ -315,6 +315,104 @@ func TestProvisionerd(t *testing.T) {
 		require.NoError(t, closer.Close())
 	})
 
+	t.Run("TemplatePlan", func(t *testing.T) {
+		t.Parallel()
+		var (
+			didComplete   atomic.Bool
+			didLog        atomic.Bool
+			didAcquireJob atomic.Bool
+			completeChan  = make(chan struct{})
+			completeOnce  sync.Once
+
+			parameterValues = []*sdkproto.ParameterValue{
+				{
+					DestinationScheme: sdkproto.ParameterDestination_PROVISIONER_VARIABLE,
+					Name:              "test_var",
+					Value:             "dean was here",
+				},
+				{
+					DestinationScheme: sdkproto.ParameterDestination_PROVISIONER_VARIABLE,
+					Name:              "test_var_2",
+					Value:             "1234",
+				},
+			}
+			metadata = &sdkproto.Provision_Metadata{}
+		)
+
+		closer := createProvisionerd(t, func(ctx context.Context) (proto.DRPCProvisionerDaemonClient, error) {
+			return createProvisionerDaemonClient(t, provisionerDaemonTestServer{
+				acquireJob: func(ctx context.Context, _ *proto.Empty) (*proto.AcquiredJob, error) {
+					if !didAcquireJob.CAS(false, true) {
+						completeOnce.Do(func() { close(completeChan) })
+						return &proto.AcquiredJob{}, nil
+					}
+
+					return &proto.AcquiredJob{
+						JobId:       "test",
+						Provisioner: "someprovisioner",
+						TemplateSourceArchive: createTar(t, map[string]string{
+							"test.txt": "content",
+						}),
+						Type: &proto.AcquiredJob_TemplatePlan_{
+							TemplatePlan: &proto.AcquiredJob_TemplatePlan{
+								ParameterValues: parameterValues,
+								Metadata:        metadata,
+							},
+						},
+					}, nil
+				},
+				updateJob: func(ctx context.Context, update *proto.UpdateJobRequest) (*proto.UpdateJobResponse, error) {
+					if len(update.Logs) != 0 {
+						didLog.Store(true)
+					}
+					return &proto.UpdateJobResponse{}, nil
+				},
+				completeJob: func(ctx context.Context, job *proto.CompletedJob) (*proto.Empty, error) {
+					didComplete.Store(true)
+					return &proto.Empty{}, nil
+				},
+			}), nil
+		}, provisionerd.Provisioners{
+			"someprovisioner": createProvisionerClient(t, provisionerTestServer{
+				provision: func(stream sdkproto.DRPCProvisioner_ProvisionStream) error {
+					err := stream.Send(&sdkproto.Provision_Response{
+						Type: &sdkproto.Provision_Response_Complete{
+							Complete: &sdkproto.Provision_Complete{
+								Resources: []*sdkproto.Resource{
+									{
+										Name: "test-resource",
+										Type: "test_type",
+										Agents: []*sdkproto.Agent{
+											{
+												Id:              "test-id",
+												Name:            "test",
+												Env:             map[string]string{},
+												StartupScript:   "echo hi",
+												OperatingSystem: "linux",
+												Architecture:    "amd64",
+												Directory:       "/home/coder",
+												Auth: &sdkproto.Agent_Token{
+													Token: "test-token",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
```

Comment:
nit, non-blocking: this is getting pretty deep, maybe a candiate to extract into some test fixtures for readability

Reply:
I removed the resources since they aren't very applicable to this test, which removes 4 levels of indentation